[5.9][CMake] Support Macros in Linux

rintaro opened the pull request:
Cherry-pick #66199 and #68082 into `release/5.9`

* **Explanation**: Enable Macros in Linux. Swift parser integration is required for enabling macros. Since Linux doesn't have ABI stable `swiftCore` in `/usr/lib` or any other stable location, executables and shared libraries that contains Swift module need to have `RUNPATH` to the runtime in its own toolchain. This PR forces "hosttools" bootstrapping mode, but set the `RUNPATH` to the currently-being-built toolchain.
* **Scope**: CMake build script
* **Risk**: Mid. This PR is carefully crafted not to affect Darwin platforms. For Linux, this is necessary for enabling macros.
* **Testing**: Passes current test suite.
* **Issues**: rdar://104346187
* **Reviewers**: TBA

DougGregor committed:
[CMake] Add paths for building Swift code against host toolchain

When building Swift code into the compiler (e.g, the new Swift parser
along with macros support), make sure we always add the appropriate
paths to (1) link against the host Swift toolchain, and (2) find the
host Swift libraries (such as SwiftSyntax) at runtime.

The CMake code for doing this was only running for Darwin builds, so
generalize it to also work on Linux.

DougGregor committed:
[CMake] Fix host library builds and rpaths for testing macros

This enables running macro tests on Linux.

DougGregor committed:
[SourceKit] Fix paths when including the new Swift compiler

rintaro committed:
[CMake] Support Macros in Linux

For compiling codes required for macro support, we now need swiftc
compiler in the build machine.

Unlike Darwin OSes, where swiftCore runtime is guaranteed to be present
in /usr/lib, Linux doesn't have ABI stability and the stdlib of the
build machine is not at the specific location. So the built compiler
cannot relies on the shared object in the toolchain.

(cherry picked from commit 9c9010e5b7506c4bd89c5ef0148e4741b32a1a2c)

rintaro committed:
Avoid adding redundant temporary RUNPATH to builder's stdlib

Many shared libs and executables are only run after stdlib/runtime are
built. They don't need to link with builders stdlib at all.

(cherry picked from commit 2a2787b2d0cdd45486e98a7cf19da4208b9ebb61)

rintaro committed:
[CMake] Fatal error when bootstrapping is not "hosttools"

and Swift parser integration is enabled.
If swift parser integration is enabled, SwiftSyntax libraries are always
built with host tools. Other SwiftCompilerSources modules must use the
same runtime with parser libraries.

(cherry picked from commit 0f0c4925ae2a70c635f4ad881991d937658b5b63)

rintaro committed:
[CMake] Update for review comments

(cherry picked from commit 9017ef51ce14f478adf2ec2640a24f3d7ce40aee)

rintaro committed:
[CMake] Fix weiredness in swift runtime linking for SourceKit libraries

SourceKit components are never built while bootstrapping stages. So it
doen't need to have stage 0 or 1 handling.

(cherry picked from commit 914a6fdf7aad9927018940b2c806f1492a4fc938)

rintaro committed:
[CMake] Further cleanup for RUNPATH setting

`$ORIGIN` is only supported in Linux, BSD, etc. (i.e. not in Windows)

(cherry picked from commit 602604be7416197cbb412972a6db5bd90d70764e)

rintaro committed:
[CMake] Set LD_LIBRARY_PATH to builder's runtime when building stdlib

In Linux. Instead of setting temporary "fallback" RUNPATH, Set
LD_LIBRARY_PATH to builder's runtime when building standard library.
So we don't need to strip the temporary RUNPATH when installing.

(cherry picked from commit 757aaa347d288c98c2514c95e025e21c2e149f53)

rintaro committed:
[CMake] Fix install for macOS toolchain

recursive-lipo requires shared libraries have +x permissions.
Use PROGRAMS instead of FILES when installing swift-syntax libraries.

(cherry picked from commit 319b8bd691d25c0a52664002efad268eec944cff)

rintaro committed:
[CMake] Fix RUNPATH of lib_InternalSwiftScan in Linux

libSwiftScan is built in 'lib' but installed in 'lib/swift/host' RUNPATH
should have correct '$ORIGIN/../{platform}' to load 'swiftCore' runtime
library.

(cherry picked from commit 4ccc86661208d2874d89fdd83030f9f6ab0a4b39)

rintaro committed:
[CMake] Stop some exectuable link with swiftCore

swift-compatibility-symbols, swift-def-to-strings-converter,
and swift-serialize-diagnostics don't use any Swift modules. But when
SWIFT_SWIFT_PARSER was enabled, they are linked with swiftCore. But
these binaries can be executed before the runtime is being built.
We need to stop them linking with swiftCore.

(cherry picked from commit dc6877334fb7bf33121fac75166313b25eebb600)

rintaro committed:
[CMake] Use RPATH_CHANGE instead RPATH_SET

RPATH_SET is not available until cmake 3.21.0. Use RPATH_CHANGE instead.

(cherry picked from commit f64506976aac4e611641415fb1c43e7f4e632d3e)

rintaro committed:
[CMake] Instantiate missing C++ templates

rintaro committed:
[CMake] Always use builder's runtime for building Swift code in Linux

Some non-stdlib thing e.g. swift-backtrace still might be built before
the runtime is built. For building Swift code in Linux "hosttools",
always set 'LD_LIBRARY_PATH' to the runtime in the builder.

(cherry picked from commit 0aa0aac72e923b53801b32bb09fc1149185a256f)

rintaro commented:
https://github.com/apple/llvm-project/pull/7345
https://github.com/apple/swift-syntax/pull/2114
@swift-ci Please Build Toolchain

bnbarham commented on cmake/modules/SwiftUtils.cmake:
```diff
@@ -107,6 +107,16 @@ function(get_bootstrapping_swift_lib_dir bs_lib_dir bootstrapping)
       get_bootstrapping_path(bs_lib_dir ${lib_dir} "1")
     endif()
+  elseif(BOOTSTRAPPING_MODE STREQUAL "HOSTTOOLS")
+    if(SWIFT_HOST_VARIANT_SDK MATCHES "LINUX|ANDROID|OPENBSD|FREEBSD")
```
There's a bunch of places we're only checking for Linux and others where we check for DARWIN and then else. Given we haven't checked the other platforms, should we just check for LINUX for now?

bnbarham commented on lib/SILOptimizer/Utils/BasicBlockOptUtils.cpp:
```diff
@@ -367,3 +367,12 @@ StaticInitCloner::clone(SingleValueInstruction *initVal) {
   return cast<SingleValueInstruction>(getMappedValue(initVal));
 }
+
+// START: Workaround for C++ interop
```
Did you pick a random file or is this the one I randomly picked? Would probably be better to put this in its own file, but given it's only needed for 5.9 ... 🤔

bnbarham committed:
[stdlib] Set PYTHONIOENCODING when building the stdlib

Ubuntu 18.04 seems to be defaulting to 'ascii', causing an error when
using line-directive combined with the utf-8 diagnostic output used when
early swift syntax is enabled.

(cherry picked from commit afd3f172cbd3c4a9a1bbe8ba7d15efbf65e754f0)

rintaro committed:
[CMake][SourceKit] Add RPATH to 'lib/swift/host'

RPATH to 'lib/swift/host' was missing in Darwin platform.

(cherry picked from commit 9d78971b9beae34bd5567ab1685ac785e6689f1f)

rintaro committed:
[CMake] Update host platform check for macro support

"Support Macros in Linux" patches was inconsistent with checking
platforms. Some only checked 'LINUX' but some matches
'LINUX|ANDROID|OPENBSD|FREEBSD'. Although I don't have tested other
platoforms than Linux at all, there's no reason to limit it to Linux.
So use the consistent check to match 'LINUX|ANDROID|OPENBSD|FREEBSD'

(cherry picked from commit 8049922861e80b79b718db54a35f5f97185a11c7)

rintaro committed:
[CMake] Handle cases where Swift parser integration is disabled in Linux

When Swift parser integration is disabled (i.e. NOT SWIFT_SWIFT_PARSER),
macro plugins targets are not created. So CMake should bail out.

(cherry picked from commit ad2dacd404c9e2994b9ae925d69101aafff3450d)

rintaro commented:
https://github.com/apple/llvm-project/pull/7345
https://github.com/apple/swift-syntax/pull/2114
@swift-ci Please test

rintaro commented:
https://github.com/apple/llvm-project/pull/7345
https://github.com/apple/swift-syntax/pull/2114
@swift-ci Please Build Toolchain

rintaro commented:
https://github.com/apple/llvm-project/pull/7345
https://github.com/apple/swift-syntax/pull/2114
@swift-ci Please test

rintaro commented:
https://github.com/apple/llvm-project/pull/7345
https://github.com/apple/swift-syntax/pull/2114
@swift-ci Please Build Toolchain

rintaro commented:
https://github.com/apple/llvm-project/pull/7345
https://github.com/apple/swift-syntax/pull/2114
@swift-ci Please test

rintaro commented:
https://github.com/apple/llvm-project/pull/7345
https://github.com/apple/swift-syntax/pull/2114
@swift-ci Please Build Toolchain
